enterprise/internal/codeintel/stores/dbstore/janitor.go

```diff
@@ -0,0 +1,175 @@
+		var commit string
+		if err := rows.Scan(&repositoryID, &repositoryName, &commit); err != nil {
+			return nil, err
+		}
+
+		sourcedCommitsMap[repositoryID] = SourcedCommits{
+			RepositoryID:   repositoryID,
+			RepositoryName: repositoryName,
+			Commits:        append(sourcedCommitsMap[repositoryID].Commits, commit),
+		}
+	}
+
+	flattened := make([]SourcedCommits, 0, len(sourcedCommitsMap))
+	for _, sourcedCommits := range sourcedCommitsMap {
+		sort.Strings(sourcedCommits.Commits)
+		flattened = append(flattened, sourcedCommits)
+	}
+
+	sort.Slice(flattened, func(i, j int) bool {
+		return flattened[i].RepositoryID < flattened[j].RepositoryID
+	})
+	return flattened, nil
+}
+
+// StaleSourcedCommits returns a set of commits attached to repositories that have been
+// least recently checked for resolvability via gitserver. We do this periodically in
+// order to determine which records in the database are unreachable by normal query
+// paths and clean up that occupied (but useless) space. The output is of this method is
+// ordered by repository ID then by commit.
+func (s *Store) StaleSourcedCommits(ctx context.Context, minimumTimeSinceLastCheck time.Duration, limit int, now time.Time) (_ []SourcedCommits, err error) {
+	ctx, traceLog, endObservation := s.operations.staleSourcedCommits.WithAndLogger(ctx, &err, observation.Args{})
+	defer endObservation(1, observation.Args{})
+
+	now = now.UTC()
+	interval := int(minimumTimeSinceLastCheck / time.Second)
+	uploadSubquery := sqlf.Sprintf(staleSourcedCommitsSubquery, sqlf.Sprintf("lsif_uploads"), now, interval)
+	indexesSubquery := sqlf.Sprintf(staleSourcedCommitsSubquery, sqlf.Sprintf("lsif_indexes"), now, interval)
+
+	sourcedCommits, err := ScanSourcedCommits(s.Store.Query(ctx, sqlf.Sprintf(staleSourcedCommitsQuery, uploadSubquery, indexesSubquery, limit)))
+	if err != nil {
+		return nil, err
+	}
+
+	numCommits := 0
+	for _, commits := range sourcedCommits {
+		numCommits += len(commits.Commits)
+	}
+	traceLog(
+		log.Int("numRepositories", len(sourcedCommits)),
+		log.Int("numCommits", numCommits),
+	)
+
+	return sourcedCommits, nil
+}
+
+const staleSourcedCommitsQuery = `
+-- source: enterprise/internal/codeintel/stores/dbstore/janitor.go:StaleSourcedCommits
+WITH candidates AS (%s UNION %s)
+SELECT r.id, r.name, c.commit
+FROM candidates c
+JOIN repo r ON r.id = c.repository_id
+-- Order results so that the repositories with the commits that have been updated
+-- the least frequently come first. Once a number of commits are processed from a
+-- given repository the ordering may change.
+ORDER BY MIN(c.max_last_checked_at) OVER (PARTITION BY c.repository_id), c.commit
+LIMIT %s
+`
+
+const staleSourcedCommitsSubquery = `
+SELECT
+	repository_id,
+	commit,
+	-- Keep track of the most recent update of this commit that we know about
+	-- as any earlier dates for the same repository and commit pair carry no
+	-- useful information.
+	max(commit_last_checked_at) as max_last_checked_at
```

Comment:
The most useless comment ever, delivered right to your inbox

```suggestion
	MAX(commit_last_checked_at) as max_last_checked_at
```

Reply:
Hey, we did this PR _together_.